packages/iov-bns/src/decode.ts

```diff
@@ -82,10 +97,35 @@ export function decodeUsernameNft(nft: codecImpl.username.IUsernameToken): BnsUs
   };
 }
 
+export function decodeNonce(acct: codecImpl.sigs.IUserData & Keyed): BcpNonce {
+  return {
+    address: encodeBnsAddress(acct._id),
+    nonce: asInt53(acct.sequence) as Nonce,
+    pubkey: decodePubkey(ensure(acct.pubkey)),
+  };
+}
+
+export function decodeToken(data: codecImpl.namecoin.IToken & Keyed): BcpTicker {
+  return {
+    tokenTicker: Encoding.fromAscii(data._id) as TokenTicker,
+    tokenName: ensure(data.name),
+  };
+}
+
 export function decodeAmount(coin: codecImpl.x.ICoin): Amount {
+  const fractionalDigits = 9; // fixed for all tickers in BNS
+
+  const wholeNumber = asNumber(coin.whole);
+  const fractionalNumber = asNumber(coin.fractional);
+
+  const quantity = Long.fromNumber(wholeNumber)
```

Comment:
This one can have errors if whole number is greater than 10^7, which is actually feasible (okay, someone moved 20% of the IOV tokens, at 8*10^7.... this will overflow. Please add a test case with such large number. (In fact starting balance of IOV reserve multisig may be close to that)

I am good with number (or Long) for whole and fractional independently, but combining them please use string operations. I do some similar string manipulations in the wallet, which aren't too hard: https://github.com/iov-one/wallet-demo/blob/master/src/logic/balances.ts (these are a bit different than your format, but it should only be a few lines to do so.

Reply:
Nice catch. I will use this docs as the max amount for tests am implementation, okay? https://github.com/iov-one/weave/blob/master/x/codec.proto#L15-L19

Do we need support for negative amounts?

Reply:
yes protobuf comment is proper range.

We use negative amounts internally in weave to perform calculations  but don't allow it in any stored data or transactions